File: src/clawcu/a2a/sidecar_plugin/openclaw/sidecar/bootstrap.js
"use strict";

// Auto-wire the a2a MCP entry into the OpenClaw config file on sidecar
// start. Merges (A2A_ENABLED=true) or removes (otherwise) `mcp.servers.a2a`
// from the JSON config located at $A2A_SERVICE_MCP_CONFIG_PATH. Safe by
// construction: any failure (missing path env, unreadable file, parse
// error) logs a warning and returns without touching the file.
//
// The caller wires this in `server.js::main()` before `server.listen()` so
// the service reads the merged config on its first MCP-config load.

const fs = require("node:fs");
const path = require("node:path");

const MCP_ENTRY_NAME = "a2a";

function _deepGet(obj, keys) {
  let cur = obj;
  for (const k of keys) {
    if (cur == null || typeof cur !== "object") return undefined;
    cur = cur[k];
  }
  return cur;
}

function _ensureObject(parent, key) {
  if (parent[key] == null || typeof parent[key] !== "object" || Array.isArray(parent[key])) {
    parent[key] = {};
  }
  return parent[key];
}

function buildMcpUrl({ port }) {
  return `http://127.0.0.1:${port}/mcp`;
}

function planBootstrap({ enabled, config, url }) {
  const safeConfig = config && typeof config === "object" && !Array.isArray(config) ? config : {};
  const current = _deepGet(safeConfig, ["mcp", "servers", MCP_ENTRY_NAME]);

  if (enabled) {
    const desired = { url };
    const same =
      current &&
      typeof current === "object" &&
      !Array.isArray(current) &&
      current.url === desired.url &&
      Object.keys(current).length === 1;
    if (same) return { action: "noop", reason: "already-present", config: safeConfig };
    const next = JSON.parse(JSON.stringify(safeConfig));
    const mcp = _ensureObject(next, "mcp");
    const servers = _ensureObject(mcp, "servers");
    servers[MCP_ENTRY_NAME] = desired;
    return { action: "merge", config: next };
  }

  if (!current) return { action: "noop", reason: "absent", config: safeConfig };
  const next = JSON.parse(JSON.stringify(safeConfig));
  const servers = _deepGet(next, ["mcp", "servers"]);
  if (servers && typeof servers === "object") {
    delete servers[MCP_ENTRY_NAME];
  }
  return { action: "remove", config: next };
}

function _atomicWriteJson(filePath, obj) {
  const tmp = `${filePath}.a2a-bootstrap.${process.pid}.tmp`;
  const text = JSON.stringify(obj, null, 2) + "\n";
  fs.writeFileSync(tmp, text, { encoding: "utf-8" });
  fs.renameSync(tmp, filePath);
}

function runBootstrap({ env, logger = console }) {
  const e = env || process.env;
  const configPath = e.A2A_SERVICE_MCP_CONFIG_PATH;
  if (!configPath) {
    logger.log("[sidecar:bootstrap] A2A_SERVICE_MCP_CONFIG_PATH unset — skipping MCP auto-wire");
    return { ok: true, action: "skip", reason: "no-config-path" };
  }
  const format = (e.A2A_SERVICE_MCP_CONFIG_FORMAT || "json").toLowerCase();
  if (format !== "json") {
    logger.log(
      `[sidecar:bootstrap] unsupported config format "${format}" — skipping (Node bootstrap handles JSON only)`,
    );
    return { ok: true, action: "skip", reason: "unsupported-format" };
  }
  const enabled = String(e.A2A_ENABLED || "").toLowerCase() === "true";
  const port = Number(e.A2A_SIDECAR_PORT || e.A2A_BIND_PORT);
  if (enabled && !(Number.isFinite(port) && port > 0)) {
    logger.warn(
      "[sidecar:bootstrap] A2A_ENABLED=true but sidecar port is unknown — skipping MCP auto-wire",
    );
    return { ok: true, action: "skip", reason: "no-port" };
  }
  const url = enabled ? buildMcpUrl({ port }) : null;

  let text;
  try {
    text = fs.readFileSync(configPath, "utf-8");
  } catch (err) {
    if (err && err.code === "ENOENT") {
      if (!enabled) {
        return { ok: true, action: "skip", reason: "file-absent-disabled" };
      }
      const next = { mcp: { servers: { [MCP_ENTRY_NAME]: { url } } } };
      try {
        fs.mkdirSync(path.dirname(configPath), { recursive: true });
        _atomicWriteJson(configPath, next);
        logger.log(`[sidecar:bootstrap] created ${configPath} with a2a MCP entry → ${url}`);
        return { ok: true, action: "create", path: configPath };
      } catch (writeErr) {
        logger.warn(`[sidecar:bootstrap] failed to create ${configPath}: ${writeErr.message}`);
        return { ok: false, action: "error", error: writeErr.message };
      }
    }
    logger.warn(`[sidecar:bootstrap] cannot read ${configPath}: ${err.message}`);
    return { ok: false, action: "error", error: err.message };
  }

  let config;
  try {
    config = text.trim() === "" ? {} : JSON.parse(text);
  } catch (err) {
    logger.warn(
      `[sidecar:bootstrap] ${configPath} is not valid JSON — refusing to overwrite (${err.message})`,
    );
    return { ok: false, action: "error", error: err.message };
  }

  const plan = planBootstrap({ enabled, config, url });
  if (plan.action === "noop") {
    logger.log(`[sidecar:bootstrap] ${configPath} already in desired state (${plan.reason})`);
    return { ok: true, action: "noop", reason: plan.reason };
  }

  try {
    _atomicWriteJson(configPath, plan.config);
    logger.log(
      `[sidecar:bootstrap] ${plan.action} a2a MCP entry in ${configPath}` +
        (url ? ` → ${url}` : ""),
    );
    return { ok: true, action: plan.action, path: configPath };
  } catch (err) {
    logger.warn(`[sidecar:bootstrap] write failed for ${configPath}: ${err.message}`);
    return { ok: false, action: "error", error: err.message };
  }
}

module.exports = {
  MCP_ENTRY_NAME,
  buildMcpUrl,
  planBootstrap,
  runBootstrap,
};

File: src/clawcu/a2a/sidecar_plugin/openclaw/sidecar/bootstrap.py
"""Auto-wire mcp.servers.a2a into openclaw.json (Python port of bootstrap.js).

Runs in server.main() before the HTTP listener starts. Merges or removes the
a2a MCP entry based on A2A_ENABLED. Refuses to overwrite invalid JSON; all
failures log a warning and return gracefully.
"""
from __future__ import annotations

import copy
import json
import os
from typing import Mapping, Optional

MCP_ENTRY_NAME = "a2a"


def _deep_get(obj, keys):
    cur = obj
    for k in keys:
        if not isinstance(cur, dict):
            return None
        cur = cur.get(k)
    return cur


def _ensure_dict(parent: dict, key: str) -> dict:
    val = parent.get(key)
    if not isinstance(val, dict):
        parent[key] = {}
    return parent[key]


def build_mcp_url(port: int) -> str:
    return f"http://127.0.0.1:{port}/mcp"


def plan_bootstrap(enabled: bool, config, url: Optional[str]):
    safe_config = config if isinstance(config, dict) else {}
    current = _deep_get(safe_config, ["mcp", "servers", MCP_ENTRY_NAME])

    if enabled:
        desired = {"url": url}
        same = (
            isinstance(current, dict)
            and current.get("url") == desired["url"]
            and len(current) == 1
        )
        if same:
            return {"action": "noop", "reason": "already-present", "config": safe_config}
        nxt = copy.deepcopy(safe_config)
        mcp = _ensure_dict(nxt, "mcp")
        servers = _ensure_dict(mcp, "servers")
        servers[MCP_ENTRY_NAME] = desired
        return {"action": "merge", "config": nxt}

    if current is None:
        return {"action": "noop", "reason": "absent", "config": safe_config}
    nxt = copy.deepcopy(safe_config)
    servers = _deep_get(nxt, ["mcp", "servers"])
    if isinstance(servers, dict):
        servers.pop(MCP_ENTRY_NAME, None)
    return {"action": "remove", "config": nxt}


def _atomic_write_json(file_path: str, obj) -> None:
    tmp = f"{file_path}.a2a-bootstrap.{os.getpid()}.tmp"
    text = json.dumps(obj, indent=2, ensure_ascii=False) + "\n"
    with open(tmp, "w", encoding="utf-8") as fh:
        fh.write(text)
    os.replace(tmp, file_path)


def run_bootstrap(env: Optional[Mapping[str, str]] = None, logger=None) -> dict:
    e = env if env is not None else os.environ
    log = logger
    if log is None:
        # Fallback to the module default_logger lazily to avoid import cycles.
        from .logsink import default_logger as _dl

        log = _dl

    config_path = e.get("A2A_SERVICE_MCP_CONFIG_PATH")
    if not config_path:
        log.info("[sidecar:bootstrap] A2A_SERVICE_MCP_CONFIG_PATH unset — skipping MCP auto-wire")
        return {"ok": True, "action": "skip", "reason": "no-config-path"}

    fmt = (e.get("A2A_SERVICE_MCP_CONFIG_FORMAT") or "json").lower()
    if fmt != "json":
        log.info(
            f"[sidecar:bootstrap] unsupported config format \"{fmt}\" — skipping (Python bootstrap handles JSON only)"
        )
        return {"ok": True, "action": "skip", "reason": "unsupported-format"}

    enabled = str(e.get("A2A_ENABLED") or "").lower() == "true"

    raw_port = e.get("A2A_SIDECAR_PORT") or e.get("A2A_BIND_PORT")
    port = None
    try:
        if raw_port not in (None, ""):
            p = int(raw_port)
            if p > 0:
                port = p
    except (TypeError, ValueError):
        port = None

    if enabled and port is None:
        log.warn(
            "[sidecar:bootstrap] A2A_ENABLED=true but sidecar port is unknown — skipping MCP auto-wire"
        )
        return {"ok": True, "action": "skip", "reason": "no-port"}

    url = build_mcp_url(port) if enabled else None

    try:
        with open(config_path, "r", encoding="utf-8") as fh:
            text = fh.read()
    except FileNotFoundError:
        if not enabled:
            return {"ok": True, "action": "skip", "reason": "file-absent-disabled"}
        nxt = {"mcp": {"servers": {MCP_ENTRY_NAME: {"url": url}}}}
        try:
            os.makedirs(os.path.dirname(config_path), exist_ok=True)
            _atomic_write_json(config_path, nxt)
            log.info(f"[sidecar:bootstrap] created {config_path} with a2a MCP entry → {url}")
            return {"ok": True, "action": "create", "path": config_path}
        except OSError as write_err:
            log.warn(f"[sidecar:bootstrap] failed to create {config_path}: {write_err}")
            return {"ok": False, "action": "error", "error": str(write_err)}
    except OSError as err:
        log.warn(f"[sidecar:bootstrap] cannot read {config_path}: {err}")
        return {"ok": False, "action": "error", "error": str(err)}

    try:
        config = {} if text.strip() == "" else json.loads(text)
    except Exception as err:
        log.warn(
            f"[sidecar:bootstrap] {config_path} is not valid JSON — refusing to overwrite ({err})"
        )
        return {"ok": False, "action": "error", "error": str(err)}

    plan = plan_bootstrap(enabled, config, url)
    if plan["action"] == "noop":
        log.info(f"[sidecar:bootstrap] {config_path} already in desired state ({plan['reason']})")
        return {"ok": True, "action": "noop", "reason": plan["reason"]}

    try:
        _atomic_write_json(config_path, plan["config"])
        suffix = f" → {url}" if url else ""
        log.info(f"[sidecar:bootstrap] {plan['action']} a2a MCP entry in {config_path}{suffix}")
        return {"ok": True, "action": plan["action"], "path": config_path}
    except OSError as err:
        log.warn(f"[sidecar:bootstrap] write failed for {config_path}: {err}")
        return {"ok": False, "action": "error", "error": str(err)}
